src/main/kotlin/uk/gov/justice/digital/hmpps/prisonercontactregistry/service/PrisonerContactRegistryService.kt

```diff
@@ -1,45 +1,94 @@
+import uk.gov.justice.digital.hmpps.prisonercontactregistry.dto.Contact
 import java.util.function.Supplier
 
 @Service
-class PrisonerContactRegistryService {
-
-  fun getPingMsg(): String {
-    return "PONG from service!"
-  }
-
-  fun getContactById(nomisPersonId: String): ContactDTO {
-    // TODO: ContactDTO -> aggregate dto or update to PrisonAPI, just stub data for now
-    // look at uk.gov.justice.digital.hmpps.pcms.service.PrisonerProfileService#getProfile
-
-    return ContactDTO.Builder(nomisPersonId)
-      .firstName("Phil")
-      .lastName("Milne")
-      .dateOfBirth(LocalDate.now())
-      .contactType("Simple Type")
-      .contactTypeDescription("Mock Contact Data")
-      .restrictions(
-        listOf(
-          RestrictionDto.Builder()
-            .restrictionType("Basic Type")
-            .restrictionTypeDescription("Mock Restriction Data")
-            .startDate(LocalDate.now())
-            .build()
-        )
-      )
-      .build()
+class PrisonerContactRegistryService(private val prisonApiClient: PrisonApiClient) {
+
+  fun getContactList(prisonerId: String, contactType: String? = null, personId: Long? = null): List<Contact> {
+    // Prisoners (Offenders) have a subset of Contacts (Persons / Visitors) which can be filtered by type and person id.
+    // When filtering the offenders contacts by type of person this results in a query result (list of matching contacts
+    // or empty list)
+    // If a prisoner is not found this results in a PrisonerNotFoundException (404).
+
+    var contacts = getContactById(prisonerId) ?: emptyList()
+
+    if (personId != null) {
+      contacts = filterByPersonId(contacts, personId)
+    }
+
+    if (!contactType.isNullOrBlank()) {
+      contacts = filterByContactType(contacts, contactType)
+    }
+
+    contacts.forEach {
+      try {
+        it.addresses = getAddressesById(it.personId)
+      } catch (e: PersonNotFoundException) {
+        // Nomis data quality issue - treat as no address data available
+        log.warn("Person not found for prisoner $prisonerId contact ${it.personId}")
+        it.addresses = emptyList()
+      }
+    }
+
+    return contacts
+  }
+
+  @Throws(PrisonerNotFoundException::class)
+  protected fun getContactById(id: String): List<Contact>? {
+    try {
+      return prisonApiClient.getOffenderContacts(id)!!.offenderContacts
+    } catch (e: WebClientResponseException) {
+      if (e.statusCode == HttpStatus.NOT_FOUND)
+        throw PrisonerNotFoundException(e.message, e)
+      throw e
+    }
+  }
+
+  @Throws(PersonNotFoundException::class)
+  protected fun getAddressesById(id: Long): List<Address> {
+    try {
+      return prisonApiClient.getPersonAddress(id)!!
+    } catch (e: WebClientResponseException) {
+      if (e.statusCode == HttpStatus.NOT_FOUND)
+        throw PersonNotFoundException(e.message, e)
+      throw e
+    }
+  }
+
+  protected fun filterByPersonId(contacts: List<Contact>?, personId: Long): List<Contact> {
```

Comment:
I think these `protected` functions can all be made private

Reply:
Well spotted thank you